pallene/assignment_conversion.lua

```diff
@@ -143,27 +148,69 @@ function Converter:apply_transformations()
                 { value = decl._type }
             )
             self:add_box_type(decl.loc, typ)
-            decl._type = typ
 
             local init_exp_update = self.update_init_exp_of_decl[decl]
 
             if init_exp_update then
                 local old_exp = init_exp_update.node
+                decl._type = typ
                 local update  = init_exp_update.update_fn
 
                 local new_node = ast.Exp.InitList(old_exp.loc, {{ name = "value", exp = old_exp }})
                 new_node._type = typ
                 update(new_node)
+
+            elseif self.lambda_of_param[decl] then
+                --- Function parameters that are captured are implementing by "proxy"-ing them.
+                --- Consider the following function that returns a closure:
+                --- ```
+                --- function m.foo(n: integer)
+                ---   -- capture n
+                --- end
+                --- ```
+                --- Since we cannot transform the declaration node of a function parameter,
+                --- we create a variable to represent the boxed parameter which can be captured.
+                --- ```
+                --- function m.foo(n: integer)
+                ---   local $n: $T = { value = n }
+                ---   -- capture and mutate $n
+                --- end
+                local param = ast.Exp.Var(decl.loc, ast.Var.Name(decl.loc, decl.name))
+                param.var._def = checker.Def.Variable(decl)
+                param.var._type = decl._type
+                param._type = decl._type
+                local decl_lhs = ast.Decl.Decl(decl.loc, "$"..decl.name, typ)
```

Comment:
If we provide a type here, it should be an `ast.Type`, not a `types.T`. We might be able to get away with setting it to `false` though.

Reply:
I think I will use the box type returned by `add_box_type`, that should be fine too I imagine.

Reply:
Since we're past the type checker, I think it's OK to not add the `ast.Type`. As long as the `_type` is there, everything should still work.